src/graph.jl

```diff
@@ -0,0 +1,349 @@
+            push!(synth_edges_df, synth_edge_attrs)
+            
+        else
+            continue
+        end
+    end
+    
+    open(output_path, "w") do io
+        print(io, repr(load_graph_data(synth_edges_df)))
+    end
+
+    #CSV.write(output_path, synth_edges_df)
+    @info "Synthetic edge dataframe generated and saved as a Julia file."
+    
+    return DataFrame(synth_edges_df)
+
+end
+
+function insert_vertices_from_jl(vertices_file::String, input_graph::Union{MetaDiGraph, Nothing})
+    
+    # Ingest data frame containing vertex information; each row represents a vertex name and type
+    # vertex_id is computed upon insertion via a hash function, and is assumed to be unique 
+    # The tuple {vertex_name, vertex_type} is assumed to be unique 
+
+    v_df = DataFrame(include(vertices_file))
+    
+    # Case 1: the graph is currently empty; objective is to create it from the input vertex and edge files.
+    if isa(input_graph, Nothing)
+        
+        n_vertices = size(v_df,1)
+        G = MetaDiGraph(n_vertices)
+        counter = 1::Int64
+
+        set_indexing_prop!(G, :v_hash)
+        
+        for v in eachrow(v_df)
+        
+            v_hash = hash(string(v.v_name, v.v_type)) 
+            
+            v_attrs = Dict(
+                :v_name=>v.v_name,
+                :v_type=>v.v_type,
+                :used=>true)
+            
+            set_indexing_prop!(G, counter, :v_hash, "$v_hash")
+            set_props!(G, counter, v_attrs)
+            @info "Inserting new vertex into graph: $v"
+            counter += 1
+        end
+    
+    # Case 2: the graph already exists (as a graph object); objective is to add new vertices/edges, and increment weights when existing edges are encountered.
+    else
+        
+        g_input = copy(input_graph)
+        n_vertices_input = nv(g_input)
+        n_vertices = size(v_df,1) + n_vertices_input
+        
+        counter = nv(g_input) + 1
+
+        G = MetaDiGraph(n_vertices)
+        
+        # insert vertices (with properties) from the input graph
+        for i in 1:nv(g_input)
+            
+            v_props = props(g_input, i)
+            
+            set_indexing_prop!(G, i, :v_hash, v_props[:v_hash])
+            
+            set_props!(G, i, 
+                Dict(:v_name=>v_props[:v_name], 
+                    :v_type=>v_props[:v_type],
+                    :used=>true))
+        end
+        
+        @info "$n_vertices_input from input graph have been inserted."
+        
+        for v in eachrow(v_df)
+        
+            v_hash = hash(string(v.v_name, v.v_type)) 
+
+            try 
+                G["$v_hash", :v_hash]
+                
+                @info "A vertex with this hash value already exists"
+                
+                # Set the "used" attribute to false, so we can remove this vertex before returning G
```

Comment:
What is the purpose of using the `used=false` part? why not just ignore these vertices?

Reply:
Or why not just use the `G["$v_hash", :v_hash]` as the preexisting vertex?

Reply:
The graph object requires an integer number of vertices when you instantiate it; I had this `false` step to handle cases where we're not sure on ingest how many unique vertices exist (e.g., whether the intersection of our existing graph and our new vertex set is a non-empty set). I figured it might be confusing/inefficient w/respect to storage if the number of vertices != number of vertices with properties.

Reply:
There is an `add_vertex!` function that will create new vertices in an existing graph. The new vertex gets added at position `nv(g) + 1` and then `nv` is incremented.

Reply:
I removed the used boolean and now just `add_vertex!` gets used.